### ISSUE TITLE
Calling api_explorer fails with 404 when required extension is set to restler_examples

### ISSUE THREAD

**Issue by dwenzel** [AOEpeople/TYPO3_Restler#7]
When setting restlers extension manager option 
`basic.extensionsWithRequiredExtLocalConfFiles = restler_examples` 
calling hostname/api_explorer/ returns:
`{
    "error": {
        "code": 404,
        "message": "Not Found"
    },
    "debug": {
        "source": "Routes.php:458 at route stage",
        "stages": {
            "success": [
                "get"
            ],
            "failure": [
                "route",
                "negotiate",
                "message"
            ]
        }
    }
}`
and an exception is logged:
`Core: Exception handler (WEB): Uncaught TYPO3 Exception: #1428562059: No restler-configuration-class found (at least one restler-configuration-class is required)! The configuration-class must be registered in ext_localconf.php of your TYPO3-extension like this: $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['restler']['restlerConfigurationClasses'][] = '[YourConfigurationClass]';The configuration-class must implement this interface: Aoe\Restler\System\Restler\ConfigurationInterface | InvalidArgumentException thrown in file /var/www/html/typo3conf/ext/restler/Classes/System/Restler/Builder.php in line 108. Requested URL: http://localhost:8333/api_explorer/`
`$GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['restler']['restlerConfigurationClasses'] = [
 0 => Aoe\Restler\System\Restler\Configuration
 1 => Aoe\RestlerExamples\System\Restler\Configuration
] `
- TYPO3 version 7.6.4
- restler version 1.2.13
- restler_examples version 0.6.0

Expected behaviour: read restler_examples' configuration and show api explorer

**Comment by soda-2005** [comment]
Hi Dirk,
the API-explorer is ONLY available, when restler-extension is also loaded (because the restler-extension 'registers' the API-explorer to restler, @see Aoe\Restler\System\Restler\Configuration::configureRestler)!

So, please try this configuration:
basic.extensionsWithRequiredExtLocalConfFiles = restler,restler_examples

If this works, than i will close the Issue.

Best regards,
Jürgen Kußmann

**Comment by dwenzel** [comment]
Hi Jürgen,
thank you for the quick reply.
I can confirm it works both with restler_examples and my custom extension key (when 'restler' is set too)
Would be nice to have a hint in the docs.

Cheers Dirk
BTW: where may I ask questions about the usage of restler? Are you on the TYPO3 mailing list?

**Comment by soda-2005** [comment]
Hi Dirk,
use this url as a starting-point to get more Informations how to develope/implement REST-endpoints with restler:
https://www.luracast.com/products/restler

There, you find links to live-examples (very helpfull) und an question&answer-section at stackoverflow.

Our TYPO3-restler-extension is so designed that it should be very simple to adapt the live-examples into the TYPO3-world.

The community of restler is very big ☺ And, no, i don't use the TYPO3-mailing list.

Cheers,
Jürgen

**Comment by dwenzel** [comment]
Thank you Jürgen, I already studied the page you mentioned. My question is specific to your extensions way to handle restler requests. 
In your examples all class properties exposed to the api are public. 
I struggled with how to re-use existing extbase models without touching their classes and could not come up with a way of mapping properties from the extbase class to new api classes.

Additionally your Loader does not load the TCA. I had to add the following code to my controller:
```php
public function __construct() 
{
  if(!$GLOBALS['TSFE'] instanceof TypoScriptFrontendController) {
    $GLOBALS['TSFE'] = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance(
      'TYPO3\\CMS\\Frontend\\Controller\\TypoScriptFrontendController',
      $GLOBALS['TYPO3_CONF_VARS'], 0, 0
    );
  }
  $GLOBALS['TSFE']->includeTCA();
  $GLOBALS['TSFE']->sys_page = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance('TYPO3\\CMS\\Frontend\\Page\\PageRepository');
}
```
in order to get Extbase' property mapping work (otherwise all properties of objects fetched from an extbase repository are NULL except pid and uid).
In short my question is: How to extend or map existing Extbase models exposing there (protected) properties via getter and setter methods?
Cheers
Dirk

**Comment by soda-2005** [comment]
Hi Dirk,
when the controller should return an object, than the properties of this object must be public, otherwise restler (the restler-API, not the restler-extension) can not 'transform' the object to an array (so restler would not be able the render the 'correct response' and would not be able to show the response-format in the API-explorer).

So, there are two options:
Option 1: Use public properties in your extbase-models
Option 2: Don't return the extbase-model in your controller. Instead of this, return an array. In this case, you would not see the data-structure of the response in the API-explorer

Here is an example, how to return an array instead of the extbase-model in your controller:

namespace Aoe\MyProductExtension\Rest\Controller;

use Aoe\MyProductExtension\Domain\Model\Product;
use Aoe\MyProductExtension\Domain\Repository\ProductRepository;
use Aoe\Restler\System\TYPO3\Loader;

class PlanController
{
    /**
     * @var ProductRepository
     */
    private $productRepository;

    /**
     * @var Loader
     */
    private $loader;

    /**
     * The TYPO3-object-manager will create this controller-object, so you can use the dependency-injection of TYPO3 here :-)
     * @param ProductRepository $productRepository
     * @param Loader $loader
     */
    public function __construct(ProductRepository $productRepository, Loader $loader)
    {
        $this->productRepository = $productRepository;
        $this->loader = $loader;
    }

    /**
     * @url GET products/{productId}
     * @param integer $productId {@min 1}
     * @return array
     */
    public function getProduct($productId)
    {
        // initialize the frontend of TYPO3 (this is required when you e.g. want to use extbase-models or if you want to render URL's in your REST-API)
        $pageIdWhichShouldBeUsedToInitializeFrontEnd = 2;
        $this->loader->initializeFrontEndRendering($pageIdWhichShouldBeUsedToInitializeFrontEnd);

        $extbaseProduct = $this->productRepository->findOne($productId);
        $restProduct = array();
        $restProduct['uid'] = $extbaseProduct->getUid();
        $restProduct['name'] = $extbaseProduct->getName();
        $restProduct['description'] = $extbaseProduct->getDescription();
        return $restProduct;
    }
}


>> Additionally your Loader does not load the TCA. I had to add the following code to my controller
Use this piece of code (in the above example) to load the TCA:
$this->loader->initializeFrontEndRendering($pageIdWhichShouldBeUsedToInitializeFrontEnd);

Another important thing when you want to use extbase-models in your REST-endpoints:
When you use the restler-extension-setting 'basic.extensionsWithRequiredExtLocalConfFiles', than you must add the extbase-extension (and all other extensions, which you maybe need during the rendering of the REST-response, e.g. realurl when you e.g. want to render realurled URL's) to this setting, e.g.:
basic.extensionsWithRequiredExtLocalConfFiles = restler,restler_examples,devlog,extbase,realurl

So, to use extbase-models in your REST-API is also very easy and simple. Maybe i should at this example to the restler_examples-extension for a better 'understanding' how to do this.

I hope, this will help you :-)

Cheers,
Jürgen
